IntegrationTests/Services/AWSSTSIntegrationTests/STSWebIdentityCredentialsProviderTests.swift

```diff
@@ -0,0 +1,206 @@
+class STSWebIdentityCredentialsProviderTests: XCTestCase {
+    private let region = "us-west-2"
+    private var oidcToken: String!
+    // File path for the cached OIDC token.
+    private var tokenFilePath: String!
+
+    // STS client with only the STS web identity credentials provider configured.
+    private var webIdentityStsClient: STSClient!
+    private var stsConfig: STSClient.STSClientConfiguration!
+
+    // Used to create identity pools and fetch cognito ID & OIDC token from it.
+    private var cognitoIdentityClient: CognitoIdentityClient!
+    // Regular STS client; used to fetch the account ID used in fetching cognito ID.
+    private var stsClient: STSClient!
+    // Cognito identity pool name & ID
+    private let identityPoolName = "aws-sts-integration-test-\(UUID().uuidString.split(separator: "-").first!.lowercased())"
+    private var identityPoolId: String!
+
+    // Used to create temporary role assumed by STS web identity credentials provider.
+    private var iamClient: IAMClient!
+    // Role properties
+    private let roleName = "aws-sts-integration-test-\(UUID().uuidString.split(separator: "-").first!.lowercased())"
+    private let roleSessionName = "aws-sts-integration-test-\(UUID().uuidString.split(separator: "-").first!.lowercased())"
+    private var roleArn: String!
+
+    // JSON assume role policy
+    private let assumeRolePolicy = """
+    {
+      "Version": "2012-10-17",
+      "Statement": [
+        {
+          "Sid": "",
+          "Effect": "Allow",
+          "Principal": {
+            "Federated": "cognito-identity.amazonaws.com"
+          },
+          "Action": [
+            "sts:AssumeRoleWithWebIdentity"
+          ],
+          "Condition": {
+             "ForAnyValue:StringLike": {
+               "cognito-identity.amazonaws.com:amr": "unauthenticated"
+             }
+          }
+        }
+      ]
+    }
+    """
+    // Identity policy name & JSON identity policy
+    private let identityPolicyName = "allow-STS-getCallerIdentity"
+    private let roleIdentityPolicy = """
+    {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Sid": "",
+                "Effect": "Allow",
+                "Action": "sts:GetCallerIdentity",
+                "Resource": "*"
+            }
+        ]
+    }
+    """
+
+    override func setUp() async throws {
+        try await super.setUp()
+
+        // Create the role to be assumed in exchange for web identity token
+        iamClient = try IAMClient(region: region)
+        let role = try await iamClient.createRole(input: CreateRoleInput(
+            assumeRolePolicyDocument: assumeRolePolicy,
+            roleName: roleName
+        ))
+        roleArn = role.role?.arn
+
+        // This wait is necessary for role creation to propagate everywhere
+        let seconds = 10
+        let NSEC_PER_SEC = 1_000_000_000
+        try await Task.sleep(nanoseconds: UInt64(seconds * NSEC_PER_SEC))
```

Comment:
We shouldn't introduce anything non-deterministic in our integration tests if we can help it. A better approach would be to poll IAM to see if the role was created successfully.

Example using `getRole`:
```
func isRoleReady(iamClient: IAMClient, roleName: String) async -> Bool {
    do {
        let response = try await iamClient.getRole(input: GetRoleInput(roleName: roleName))
        return response.role != nil
    } catch {
        return false
    }
}
```

Then you can setup a function that calls isRoleReady every X seconds using Task.sleep until a configured timeout (if after X seconds/minutes the role is still not there, throw an error).

Reply:
It definitely would be ideal to have some polling mechanism like the one you suggested instead of a fixed wait behavior, but `getRole` doesn't serve our purpose here (nor does any other IAM API). That's because changes to IAM entities become available to IAM API immediately BUT it takes time for that info to propagate everywhere. With 7 second wait it reliably passed every time on local machine, so I made it 10 just to be safe. Here's a [relevant stackoverflow discussion](https://stackoverflow.com/questions/20156043/how-long-should-i-wait-after-applying-an-aws-iam-policy-before-it-is-valid/20157294#20157294) if you'd like to read more about it.

Reply:
Also see if there is a waiter defined on the service for the resource you are waiting on.

Check the Kinesis integration test: it uses a predefined waiter and manual waiting in two stages of the test to ensure conditions are met before proceeding.

Reply:
Tried the waiters and they don't serve the polling purpose in this instance for the same reasons explained above, as expected. Unless IAM provides some sort of API where it shows IAM identity's propagation status, I don't think it's possible to poll this info.

Reply:
Okay, we can try this live in CI for a while, but we'll have to revisit this test if it causes spurious failures in practice.

With the size & complexity of our test suite, we can't afford even a small number of flaky tests.